### PR TITLE
add ivp package and subpackages for integrating initial value problems

### PR DESCRIPTION
**Pull request by soypat**
Added Runge Kutta 4th order multivariable integration scheme. Example added simulates a falling point-mass.

I expect some more changes may be requested. I'm all ears!

Checklist:

- API changes have been discussed: No API changes
- code is goformated correctly (goimports): Check
- packages with generated code have had code generation run: No code gen
- tests pass locally: Tests added pass!
- linked to relevant issues: None I'm aware of

### CODE REVIEW

File: ivp/ivp.go
Comment (kortschak) [inline ivp/ivp.go:1]: Please add appropriate copyright headers to all files.

File: ivp/ivp.go
Comment (kortschak) [inline ivp/ivp.go:5]: Single line, please.

File: ivp/sim/ivpmodel.go
Comment (kortschak) [inline ivp/sim/ivpmodel.go:1]: I don't think we need a proliferation of directories here. This can be merged with ivp/ivp.go.

File: ivp/solver/algorithm_test.go
Comment (kortschak) [inline ivp/solver/algorithm_test.go:1]: Again, this can go into ivp.

File: ivp/solver/algorithms.go
Comment (kortschak) [inline ivp/solver/algorithms.go:1]: Same.

File: ivp/sim/ivpmodel.go
Comment (kortschak) [inline ivp/sim/ivpmodel.go:13]: Please put godoc comments on exported functions.

File: ivp/solver/algorithm_test.go
Comment (kortschak) [inline ivp/solver/algorithm_test.go:45]: Put the code in-line rather than calling out to setup.

File: ivp/solver/algorithm_test.go
Comment (kortschak) [inline ivp/solver/algorithm_test.go:79]: Example tests are for teaching, so this needs more commentary to lead the user through how to use the package.

**Review by kortschak — CHANGES_REQUESTED** [review]
Thank you for the contribution.

The should have a proposal issue for discussion, but I've gone through and done an initial review and what is here can be used as a focus for discussion there.

I think the most appropriate person to involve in the discussion would be @vladimir-ch, but time availability may be a limiting.

We normally wait until approval before including the changes to A+C since there will be revisions to code in the PR and we don't want non-logic changes tangled with logic changes. So would you please reset hard to the first commit of this branch? We'll ask you to send another PR for the A+C change when the time comes.

**Comment by soypat** [comment]
@kortschak Thanks for the suggestions. I've applied them to the new commit after a hard reset.

File: ivp/ivp.go
Comment (soypat) [inline ivp/ivp.go:8]: go imports failed me. should i reset hard again?

Comment (kortschak) [reply ivp/ivp.go:8]: No, that's fine, it can be cleaned up after discussion. Do you want to open a proposal to add this package?

File: ivp/doc.go
Comment (sbinet) [inline ivp/doc.go:1]:
```suggestion
// Copyright ©2021 The Gonum Authors. All rights reserved.
```

File: ivp/ivp.go
Comment (sbinet) [inline ivp/ivp.go:1]:
```suggestion
// Copyright ©2021 The Gonum Authors. All rights reserved.
```

File: ivp/rungekutta4.go
Comment (sbinet) [inline ivp/rungekutta4.go:1]:
```suggestion
// Copyright ©2021 The Gonum Authors. All rights reserved.
```

File: ivp/rungekutta4_test.go
Comment (sbinet) [inline ivp/rungekutta4_test.go:1]:
```suggestion
// Copyright ©2021 The Gonum Authors. All rights reserved.
```

File: ivp/model.go
Comment (sbinet) [inline ivp/model.go:1]:
```suggestion
// Copyright ©2021 The Gonum Authors. All rights reserved.
```

**Review by sbinet — COMMENTED** [review]
it's great somebody is picking up the slack on this front.

I haven't looked at the PR/proposal in detail yet, but after my 10,000ft diagonal read, I wondered: wouldn't it make sense to use the same kind of mechanism/protocol for the iteration than what we used for `optimize`? would it make sense?

**Comment by kortschak** [comment]
> I wondered: wouldn't it make sense to use the same kind of mechanism/protocol for the iteration than what we used for optimize? would it make sense?

This is why I thought Vladimir would be interested. Also the relation to ODE solvers which has been on his distant radar for a long time.

**Comment by vladimir-ch** [comment]
Thanks for the contribution!

@kortschak is right, an IVP solver package has been on my distant radar for a very long time, almost from the beginning of my involvement with Gonum, see https://groups.google.com/g/gonum-dev/c/gbsTBk2xY6M/m/bxZWbXJiBAAJ. Occasionally an external nudge comes showing interest in such package, for example https://github.com/gonum/gonum/issues/532 in 2018, but getting the API and implementation right is a major undertaking which doesn't fit well with my limited availability and now in the pandemic also with limited uninterrupted time to focus.

Recently I've been focusing more on the `mat` and `lapack` packages to have the basics (a more feature-complete support for solving linear systems, more matrix types like (symmetric) banded or tridiagonal matrices, maybe even sparse matrices, iterative solvers on which I have been procrastinating for several years now, ...) in place before proceeding to these more specialized packages. My reasoning was that implicit solvers need to solve linear systems at every step so let's have the base done but now I think the work on an ode package should proceed in smaller steps and start with explicit solvers only.

If we had an elementary working Go implementation of an explicit Runge-Kutta solver like [DOPRI5](http://www.unige.ch/~hairer/prog/nonstiff/dopri5.f) in http://www.unige.ch/~hairer/software.html, that would be a great start. This particular implementation is short, just 700 lines including 200 lines of documentation, and yet has stepsize control and dense output. We could use it as a basis for an API discussion and build a test suite around it.

I think this PR is not ready for merging directly into Gonum, so I'd propose to instead submit it to [the exp repo](https://github.com/gonum/exp). I also suggest to go through those old discussions, I think that a lot of the ideas are still valid.

If you want to take on this project, my availability is better recently and so I may have some time for discussions and code review.

File: ivp/ivp.go
Comment (vladimir-ch) [inline ivp/ivp.go:36]: This looks as a good starting point but the question arises how the user will interact with the package. Will he instantiate a integrator and implement a loop to step forward? Then this interface is not needed (the user already has a concrete type). Or will this package implement the integration loop and take the integrator as this (or similar) interface?

**Comment by soypat** [comment]
@vladimir-ch I've taken your suggestions and applied some of them here: https://github.com/gonum/exp/pull/54

I'll be adding Dormand-Prince method soon if you say so.

**Comment by vladimir-ch** [comment]
> I'll be adding Dormand-Prince method soon if you say so.

I mentioned the `dopri5.f` code not because of the method itself but rather because of it being a solid, small but still fairly capable implementation of an IVP solver. Let's ignore it for now and focus on what you have.

**Pull request closed without merging**
